### PR TITLE
feat: debug options

### PR DESCRIPTION
**Pull request by pierrezimmermannbam**
### Summary

Adds optional filterProps option for debug util as discussed in #1135

### Test plan

I added some tests using the option and filtering through propName, propValue or node
Additionally I checked that filtering props in debug doesn't mutate initial instance

### CODE REVIEW

File: src/__tests__/render.test.tsx
Comment (AugustinLF) [inline src/__tests__/render.test.tsx:208]: I'd either break those in different tests (to have better snapshot tests), and/or inline the snapshots. The snapshot are very "isolated" right now

File: src/render.tsx
Comment (AugustinLF) [inline src/render.tsx:138]: Please update flow types (I wish we could drop them^^')

**Review by AugustinLF — CHANGES_REQUESTED** [review]
Looks pretty good overall (even though I'm not familiar with the way `pretty-format` works, two additional points:
- can we get `mapNode` also as an option?
- can you try adding this option to the global `configure` options? I believe there is a lot of values in setting a value for a whole project (I'll definitely filter out all my `style` props)

File: src/render.tsx
Comment (pierrezimmermannbam) [reply src/render.tsx:138]: Oh I forgot about these, I updated them

**Comment by pierrezimmermannbam** [comment]
@AugustinLF I added the option to configure the debugOptions globally and treated your other comments. Also I have updated the website docs with the new api. Regarding the mapNode option I was considering adding it in another pr if that's ok with you, so that the review process is shorter and also because I think this option by itself makes sense so the faster it is available the better

File: package.json
Comment (mdjastrzebski) [inline package.json:101]: Is this change related to his PR? Why is it needed?

File: src/config.ts
Comment (mdjastrzebski) [inline src/config.ts:6]: RTL/DTL convention seems to be [`defaultXxx`](https://testing-library.com/docs/dom-testing-library/api-configuration#defaulthidden)
```suggestion
  defaultDebugOptions?: Partial<DebugOptions>;
```

File: src/helpers/format.ts
Comment (mdjastrzebski) [inline src/helpers/format.ts:23]:
```suggestion
const reactTestComponentPlugin = (filterProps?: FilterNodeByTypeFunction): NewPlugin => {
```

File: src/helpers/format.ts
Comment (mdjastrzebski) [inline src/helpers/format.ts:19]: Seems like we might as well use the orginal ReactElement plugin and just tweak the ReactTestComponent
```suggestion
    plugins: [reactTestComponentPlugin(options.filterProps), plugins.ReactElement],
```

File: src/helpers/format.ts
Comment (mdjastrzebski) [inline src/helpers/format.ts:27]: As mentioned above you should be able to make this impl. easier by just focusing on `ReactTestComponent` part and using `plugins.ReactElement` as-is.

File: typings/index.flow.js
Comment (mdjastrzebski) [inline typings/index.flow.js:321]: nit: I would go with interface here, but this is purely personal style pref.

```suggestion
interface FormatOptions {
  filterProps?: FilterNodeByTypeFunction,
};

interface DebugOptions extends FormatOptions {
  message?: string,
}
```

File: typings/index.flow.js
Comment (mdjastrzebski) [inline typings/index.flow.js:324]:
```suggestion
  (options?: DebugOptions | string): void,
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:177]:
```suggestion
debug(options?: DebugOptions | string): void
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:187]: Let's suggest them only the modern way.
```suggestion
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:188]: formatting:
```suggestion
screen.debug({ message: 'optional message' });
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:211]:
```suggestion
This will log the rendered JSX without the `style` prop.
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:760]:
```suggestion
Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`.
```

**Review by mdjastrzebski — COMMENTED** [review]
I've done initial review on implementation and left some comments there. Overall I like the code so far, looks pretty solid. 

The point I want to rise is the desired API we want to have. 

The current API is:
```
export type FormatOptions = {
  filterProps?: (
    propName: string,
    propValue: unknown,
    node: ReactTestRendererJSON
  ) => boolean;
};
```
While it's quite flexible, I think we could investigate alternative approaches, based on the use cases we want to support.

### Idea 1: allowlist/blocklist
As far as I understand the primary use case is to print only some relevant props and/or exclude some props that are cluttering the output. We are rarely interested in inspecting the prop value. We could achieve that by somewhat easier API:

```ts
type FormatOptions = {
  allowedProps?: string[];
  ignoredProps?: string[];
}
```

If specified `allowedProps` will contain the names props that will be printed, no other props would be printed by `format`. Otherwise, all props will be printed except the ones specified in `ignoredProps`.

Pros: simpler API for majority use case
Cons: less flexible

### Idea 2: ability to transform props
Opposite to idea 1, this makes the API more flexible by allow prop transformation before printing. This could allow for both removal of props that are not interesting to as, or applying some transformation to their value that would improve the readability:

Version 1 - per prop
```ts
type FormatOptions = {
  mapProp?: (
    propName: string,
    propValue: unknown,
    node: ReactTestRendererJSON
  ) => [name, value];
}
```

Version 2 - per node
```ts
type FormatOptions = {
  mapProps?: (props: Record<string, unknown>) => Record<string, unknown>; // output props to be displayed
}
```

### Examples
Original:
```ts
debug(<JSX />, { filterProps: (name, value, node) => name === 'style' });
debug(<JSX />, { filterProps: (name, value, node) => ['style', 'testID'].includes(name) });
```

Idea 1:
```ts
debug(<JSX />, { allowedProps: ['style'] });
debug(<JSX />, { allowedProps: ['style', 'testID'] });
```

Idea 2 - per prop:
```ts
debug(<JSX />, { mapProp: (name, value, node) => name === 'style' ? [name, value] : undefined });
debug(<JSX />, { mapProp: (name, value, node) => ['style', 'testID'].includes(name ? [name, value] : undefined });
```

Idea 2 - per node:
```ts
debug(<JSX />, { mapProps: ({ style }) => ({ style }) });
debug(<JSX />, { mapProps: ({ style, testID }) => ({ style, testID }) });
```

@pierrezimmermannbam could you provide typical examples you envision this option will handle so that we can figure out which API options is the best to address it.

File: src/__tests__/__snapshots__/render.test.tsx.snap
Comment (AugustinLF) [inline src/__tests__/__snapshots__/render.test.tsx.snap:152]: Shouldn't `children` be filtered also?

**Review by AugustinLF — APPROVED** [review]
~Looks good to me, my feedback has been addressed, I'll let you address @mdjastrzebski comments~

**Review by AugustinLF — COMMENTED** [review]
Good for me, I'll let you address @mdjastrzebski's comment meanwhile. I won't hit the `Approve` button to be sure it doesn't get merged by mistake :)

**Comment by pierrezimmermannbam** [comment]
@mdjastrzebski you're right we should consider alternative approaches. I plan to use this feature in the following ways : 

- remove globally some props that aren't useful for debugging with configure api (style or path for instance)
- create custom debug utils that would be used for debugging each query using screen (for instance debugText would filter all props except children)
- use a more custom version in tests if the need arises but this should be pretty rare

Overall I'd mostly use this either with configure or have a set of utils that I would write only once so I like the idea of having a rather flexible api. However It should be accessible to users too so maybe we could document it a bit more with some use cases that we find relevant.

AllowedProps and ignoredProps is probably simpler than filterProps so I think it's interesting. I'm  not sure however that it's that much simpler, plus i'm afraid that there may be some conflicts between ignored and allowed props especially if one has been set through configure. Also i was thinking of globally filtering props whose value is a function which you cannot do without access to the propValue. I also like to have a single option for filtering props, because we may add other options later 
and the more there are the more it's going to be complicated to understand all of them.

mapProps is very interesting though, being able to flatten styles in the debug function for instance could be a very useful option. Are there any other use cases though ? Maybe directly having a flattenStyle option would be more convenient if that's not the case. What I'm afraid of is that it seems less obvious to me why you'd want to map props than why you'd want to filter props but it would indeed allow way more flexibility.  I think I wouldn't mind having both the filterProps and the mapProps option though. 

What do you think would be the best api based on the examples I provided ?

**Comment by mdjastrzebski** [comment]
@pierrezimmermannbam not sure if you're participating in Hacktoberfest but if so I've added `hacktopberfest-accept`, I plan to review this again in the following days.

File: src/__tests__/__snapshots__/render.test.tsx.snap
Comment (mdjastrzebski) [reply src/__tests__/__snapshots__/render.test.tsx.snap:152]: I think this is due to `children` prop being extracted from `props` in both TestInstance and JSON. Imo we should keep this as is, as a reasonable behaviour for most cases.

**Comment by mdjastrzebski** [comment]
@pierrezimmermannbam 

Imo based on your use cases I think `mapProps` (per node version) makes most sense and is also the most flexible approach. It allows for both filtering & transforming props (e.g. flatten styles), while also allowing for basic output on a prop combination, e.g. output `style` prop only if `disabled` props is not set, etc. I'm curious about @AugustinLF take on the API.

re accept list/ignore list priority: I think a reasonable assumption is that `accept` list takes priority over `ignore` list. If you have `accept` list then you do not use `ignore` list at all.

re ease of use: we should have some documentation examples for common use cases, that should clarify the usage.

re having both `filterProp` and `mapProps` that would make the API a bit confusing as we would have two functions doing very similar things, and also it's not clear how should they interact. Which should be used if both are present, etc

I would be wary of introducing niche options like `flattenStyles` as it can be done with `mapProps` and imo is not worth to complicating our API.

**Comment by AugustinLF** [comment]
I think it's fair to go with `mapProps`. It is an option for advanced users, so we should aim it to make it complete. And yes, for things like `flattenStyles` we could offer it as a recipe. It doesn't have to be in our API. Or if highly requested, we could expose it as an export, that could then be passed to the `mapProps`.

**Comment by pierrezimmermannbam** [comment]
I also agree, I'll change the api and add some examples of usage in the documentation

File: package.json
Comment (pierrezimmermannbam) [reply package.json:101]: It's not really related to this pr but I noticed that mocks weren't automatically cleared because I had to add tests on debug and since the calls to console.log were not cleared, my tests did not pass after I ran them with the other ones. I think tests should all pass when they are run by themselves and should not depend on other tests and automatically clearing mocks call history helps achieving that

File: src/__tests__/__snapshots__/render.test.tsx.snap
Comment (pierrezimmermannbam) [reply src/__tests__/__snapshots__/render.test.tsx.snap:152]: I was surprised by this as well but I actually think that this is fine, we'll want to keep the children prop in most cases and not having to explicitly include it is nice

**Comment by pierrezimmermannbam** [comment]
@mdjastrzebski  I changed the api to use mapProps instead and updated the website docs to include more examples of the option usage. I also treated the comments you left. 

There is one last thing I'm not sure about though, should we keep the compatibility with previous versions for debugOptions ? The debug util is not meant to stay in the code, it's just a debugging tool so there should be little to no issues regarding the migration. I feel like it adds unnecessary complexity in the RNTL codebase while providing very little value to users. I think this could be merged as is and be included in a release before next major version but I would remove retro compatibility with the next major version, wdyt ?

File: src/render.tsx
Comment (mdjastrzebski) [inline src/render.tsx:158]:
```suggestion
        'Using debug("message") is deprecated and will be removed in future release, please use debug({ message; "message" }) instead.'
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:787]: I would rather refer user to `debug` `options` docs for details rather than repeating the use cases.

```suggestion
Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`. 
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:253]: The concept of building test utils seems like a useful idea, however the `typeEmail` example seems quite artificial and I would not recommend using such util over explicit `fireEvent` + query.

@pierrezimmermannbam do you have idea for something that we could actually recommend to your users here?

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:205]:
```suggestion
You can use the `mapProps` option to transform the props that will be printed : 
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:209]: formatting:
```suggestion
render(<View style={{ backgroundColor: 'red' }}/>);
debug({ mapProps : ({style, ...props}) => ({ props }) })
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:214]: Not sure about mentioning `children` prop in this case, as it is treated specially by `mapProps`

```suggestion
This option can be used to target specific props when debugging a query (for instance keeping only `children` prop when debugging a `getByText` query) and you can also transform prop values so that they are more readable (e.g. flatten styles).
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:216]:
```suggestion
The `children` prop cannot be filtered out so the following will print all rendered components with all props but children filtered.
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:184]: Pls add lower-level headers for options as we do for other APIs

```suggestion
#### `message` option
```

File: website/docs/API.md
Comment (mdjastrzebski) [inline website/docs/API.md:204]:
```suggestion
#### `mapProps` option
```

File: typings/index.flow.js
Comment (mdjastrzebski) [inline typings/index.flow.js:429]:
```suggestion
    defaultDebugOptions?: $Shape<DebugOptions>;
```

File: typings/index.flow.js
Comment (mdjastrzebski) [inline typings/index.flow.js:363]: `FormatOptions` does not seem to be used anywhere, let's merge it with `DebugOptions` for flow


```suggestion
type DebugOptions = {
  message?: string,
  mapProps?: MapPropsFunction,
};
```

File: src/render.tsx
Comment (mdjastrzebski) [inline src/render.tsx:149]: style, nit:
Pls move `const json = ...` line just above the `if (json)` statement, as it is not used during option calculation part.

File: src/helpers/format.ts
Comment (mdjastrzebski) [inline src/helpers/format.ts:18]: I think it's still a good idea to make `getCustomPlugin` tweak only `ReactTestComponent` while passing `ReactElement` normally
```suggestion
    plugins: [getTestComponentPlugin(options.mapProps), plugins.ReactElement],
```

File: src/helpers/debugDeep.ts
Comment (mdjastrzebski) [inline src/helpers/debugDeep.ts:18]: Should prevent crash if user passes `null`, because `typeof null === 'object'`

```suggestion
    typeof options === 'object' ? { mapProps: options?.mapProps } : undefined;
```

File: src/__tests__/render.test.tsx
Comment (mdjastrzebski) [inline src/__tests__/render.test.tsx:224]: This test seems confusing, as `children` in not included in the props. Pls adjust or remove the test.

File: src/__tests__/render.test.tsx
Comment (mdjastrzebski) [inline src/__tests__/render.test.tsx:256]: Having `getByText` assertion does not seem to contribute to the test.
```suggestion
```

File: src/__tests__/render.test.tsx
Comment (mdjastrzebski) [inline src/__tests__/render.test.tsx:273]: Having `getByText` assertion does not seem to contribute to the test.
```suggestion
```

File: src/__tests__/render.test.tsx
Comment (mdjastrzebski) [inline src/__tests__/render.test.tsx:288]: Having `getByText` assertion does not seem to contribute to the test.
```suggestion
```

File: src/__tests__/render.test.tsx
Comment (mdjastrzebski) [inline src/__tests__/render.test.tsx:214]: What caused the index change here?

File: src/helpers/format.ts
Comment (mdjastrzebski) [inline src/helpers/format.ts:29]: Not sure if it's a good idea to mutate the passed `val` object. Can we rather use object spread operator to create a modified copy?

**Review by mdjastrzebski — CHANGES_REQUESTED** [review]
@pierrezimmermannbam This PR looks really solid. I've added some suggestions for finishing touches and edge cases. 

Pls add some real-life examples to the docs on how users could you this to improve their debugging output.

File: website/docs/API.md
Comment (pierrezimmermannbam) [reply website/docs/API.md:787]: agreed, I added a link to debug options doc

File: website/docs/API.md
Comment (pierrezimmermannbam) [reply website/docs/API.md:253]: This is something is sometimes do when I test forms and have a lot of tests, imo it makes tests more readable and very easy to understand while not making them harder to debug since you still have the same error messages. It's even more interesting when it allows to hide testID usage since the queries by testID do not really help to understand the test. An example of this is waiting for a loader to disapear, which is often done in every test when testing at page level

```tsx
const waitForLoaderToDisapear = async () => {
   await waitForElementToBeRemoved(() => screen.getByTestID('PageLoader'));
}
```

Does this example make more sense to you ?

File: website/docs/API.md
Comment (pierrezimmermannbam) [reply website/docs/API.md:214]: I changed the order so that it is explained first that children prop cannot be filtered then give the example of usage, since I believe this will be one of the most common, at least until byRole queries become more popular

File: website/docs/API.md
Comment (pierrezimmermannbam) [reply website/docs/API.md:184]: Very good idea I hadn't noticed this was done for other apis !

File: typings/index.flow.js
Comment (pierrezimmermannbam) [reply typings/index.flow.js:429]: Oh good catch, I keep missing the flow types updates

File: src/helpers/format.ts
Comment (pierrezimmermannbam) [reply src/helpers/format.ts:18]: Absolutely, it also makes the code way simpler !

File: src/__tests__/render.test.tsx
Comment (pierrezimmermannbam) [reply src/__tests__/render.test.tsx:224]: Oh you're right, I had written them before I realized the children prop wasn't filtered and it does not really make sense anymore. I deleted one test and modified the other one

File: src/__tests__/render.test.tsx
Comment (pierrezimmermannbam) [reply src/__tests__/render.test.tsx:256]: Actually what I was trying to do here was make sure that filtering out props wouldn't really delete them but since the children prop is never filtered it doesn't make sense. I'll replace this by a more explicit and relevant test

File: src/__tests__/render.test.tsx
Comment (pierrezimmermannbam) [reply src/__tests__/render.test.tsx:214]: The fact that mocks are being cleared automatically do to the change I made in jest.config.js. We were checking the 4th call because it was called in previous test but now mock calls are reset before each test

File: src/helpers/format.ts
Comment (pierrezimmermannbam) [reply src/helpers/format.ts:29]: I had my doubt on this, this is why I added a test to verify that the original object is not modified and it's not the case so I think it's fine as is. Also I believe that using the spread operator would create a shallow copy so it would not really be a protection against modifications either

**Comment by pierrezimmermannbam** [comment]
@mdjastrzebski  I think I answered all your comments. I added an example to the doc and detailed more the flatten style example, do you think of anything else ?

File: src/helpers/format.ts
Comment (mdjastrzebski) [reply src/helpers/format.ts]: format works on JSON output not ReactTestInstances, but still invoking `format` would mutate the JSON output. I've changed that to object copy, not mutation.

**Review by mdjastrzebski — APPROVED** [review]
Looks great! Good work @pierrezimmermannbam 🚀 

I've made some cosmetic tweaks to docs and code before merging.

**Pull request merged**
